Redis Backend Updates: WatchPrefix Support, Hash Support, Standard Redis Key Naming Convention

jmooradi opened the pull request:
I've made some updates to the redis backend based on some needs for a project I was working on. I figured some of these might be useful for others as well as I wanted a second look at all the code for additional feedback on any problems you might see in the implementation especially since I'm fairly new to golang.

Overview of additions:
- WatchPrefix support using redis pubsub and keyspace events (if you don't want to enable keyspace events on your redis db you can also manually trigger a watchprefix using publish ex: `publish __keyspace@db__:prefix:key set`)
- GetValues rewrite to support redis hashes which is preferable since a normal scan has to iterate over every key in the entire database #532
- Convert confd "/" key naming notation to redis standard ":" notation when doing key lookups, this change will break existing setups, maybe it would be better to use a config flag to enable this?

okushchenko commented on backends/redis/client.go:
```diff
@@ -110,49 +140,151 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
+		}
+	}()
+
+	for {
```
This loop is redundant. The `WatchPrefix` function exists after a return.

okushchenko commented on backends/redis/client.go:
```diff
@@ -11,17 +11,25 @@ import (
 	client   redis.Conn
 	machines []string
 	password string
+	psc redis.PubSubConn
```
Formatting.

jmooradi commented:
I added the changes you suggested and added a "separator" flag that I figured might be useful for other backends as well. -separator ":" enables the redis native notation, you can also specify any string really
